buoy_gazebo/src/MooringForce/MooringForce.cpp

```diff
@@ -0,0 +1,322 @@
+  igndbg << "Anchor position set to " << this->dataPtr->anchorPos
+    << std::endl;
+
+  this->dataPtr->effectiveRadius = _sdf->Get<double>(
+    "enable_beyond_radius", this->dataPtr->effectiveRadius).first;
+  igndbg << "Effective radius set to beyond " << this->dataPtr->effectiveRadius
+    << std::endl;
+
+  // Find necessary model links
+  if (this->dataPtr->FindLinks(_ecm)) {
+    this->dataPtr->UpdateVH(_ecm);
+  }
+
+  this->dataPtr->B.resize(1);
+}
+
+//////////////////////////////////////////////////
+void MooringForce::PreUpdate(
+  const gz::sim::UpdateInfo & _info,
+  gz::sim::EntityComponentManager & _ecm)
+{
+  GZ_PROFILE("MooringForce::PreUpdate");
+
+  // If necessary links have not been identified yet, the plugin is disabled
+  if (this->dataPtr->heaveConeLinkEnt == gz::sim::kNullEntity ||
+    this->dataPtr->buoyLinkEnt == gz::sim::kNullEntity) {
+    this->dataPtr->FindLinks(_ecm);
+    return;
+  }
+
+  // TODO(anyone): Support rewind
+  if (_info.dt < std::chrono::steady_clock::duration::zero()) {
+    ignwarn << "Detected jump back in time [" <<
+      std::chrono::duration_cast<std::chrono::seconds>(_info.dt).count() <<
+      "s]. System may not work properly." << std::endl;
+  }
+
+  // Nothing left to do if paused.
+  if (_info.paused) {
+    return;
+  }
+
+  // Update V and H based on latest buoy position
+  this->dataPtr->UpdateVH(_ecm);
+  // If buoy is not far enough from anchor for there to be a need to pull it
+  // back, no need to apply mooring force
+  if (this->dataPtr->H < this->dataPtr->effectiveRadius) {
+    return;
+  }
+
+  Eigen::HybridNonLinearSolver<CatenaryHSoln> catenarySolver(
+    this->dataPtr->catenarySoln);
+  // Tolerance for error between two consecutive iterations
+  catenarySolver.parameters.xtol = 0.0001;
+  // Max number of calls to the function
+  catenarySolver.parameters.maxfev = 1000;
+
+  // Initial guess B = L - V - b
+  double b = 1.0;
+  this->dataPtr->B[0] = this->dataPtr->L - this->dataPtr->V - b;
+
+  // Scaling factor
+  double c = CatenaryFunction::CatenaryScalingFactor(
+    this->dataPtr->V, this->dataPtr->B[0], this->dataPtr->L);
+
+  int solverInfo;
+  while ((c <= 1e-5) && (this->dataPtr->L - this->dataPtr->V - b > 0.0)) {
+    // Invalid chain length. One side of triangle is negative length.
+    if ((this->dataPtr->L - this->dataPtr->B[0]) *
+        (this->dataPtr->L - this->dataPtr->B[0]) -
+        this->dataPtr->V * this->dataPtr->V <= 0) {
+      b += 1.0;
+      continue;
+    }
+
+    // Update guess for B, with new b
+    this->dataPtr->B[0] = this->dataPtr->L - this->dataPtr->V - b;
+
+    // Solve for B, pass in initial guess
+    solverInfo = catenarySolver.solveNumericalDiff(this->dataPtr->B);
+
+    // Recalculate c with newly solved B
+    c = CatenaryFunction::CatenaryScalingFactor(
+      this->dataPtr->V, this->dataPtr->B[0], this->dataPtr->L);
+
+    // Increment b, to iterate on B
+    b += 1.0;
```

Comment:
This is a little more readable imho and matches the python code
```suggestion
  double b = 0.0;

  // Catenary scaling factor
  double c = 0.0; 

  int solverInfo;
  while ((c <= 1e-5) && (this->dataPtr->L - this->dataPtr->V - b > 0.0)) {
    // Increment b, to iterate on B
    b += 1.0;

    // Update guess for B, with new b
    this->dataPtr->B[0] = this->dataPtr->L - this->dataPtr->V - b;

    // Invalid chain length. One side of triangle is negative length.
    if ((this->dataPtr->L - this->dataPtr->B[0]) *
        (this->dataPtr->L - this->dataPtr->B[0]) -
        this->dataPtr->V * this->dataPtr->V <= 0) {
      continue;
    }

    // Solve for B, pass in initial guess
    solverInfo = catenarySolver.solveNumericalDiff(this->dataPtr->B);

    // Recalculate c with newly solved B
    c = CatenaryFunction::CatenaryScalingFactor(
      this->dataPtr->V, this->dataPtr->B[0], this->dataPtr->L);
```